File: samples/da-trey-research-python/src/services/consultant_api_service.py
from datetime import datetime

from ..model.base_model import Consultant, HoursEntry
from ..model.api_model import ApiConsultant, ApiConsultantAssignment, ApiChargeTimeResponse
from .project_db_service import project_db_service
from .assignment_db_service import assignment_db_service
from .consultant_db_service import consultant_db_service
from .utilities import HttpError

AVAILABLE_HOURS_PER_MONTH = 160


class ConsultantApiService:
    """API service for consultant operations."""

    def get_api_consultant_by_id(self, consultant_id: str) -> ApiConsultant:
        consultant = consultant_db_service.get_consultant_by_id(consultant_id)
        if consultant:
            assignments = assignment_db_service.get_assignments()
            return self._get_api_consultant_for_base(consultant, assignments)
        return None

    def get_api_consultants(
        self, consultant_name: str, project_name: str, skill: str,
        certification: str, role: str, hours_available: str
    ) -> list:
        consultants = consultant_db_service.get_consultants()
        assignments = assignment_db_service.get_assignments()

        # Filter on base properties
        if consultant_name:
            consultants = [c for c in consultants if consultant_name.lower() in c.name.lower()]
        if skill:
            consultants = [
                c for c in consultants
                if any(skill.lower() in s.lower() for s in c.skills)
            ]
        if certification:
            consultants = [
                c for c in consultants
                if any(certification.lower() in cert.lower() for cert in c.certifications)
            ]
        if role:
            consultants = [
                c for c in consultants
                if any(role.lower() in r.lower() for r in c.roles)
            ]

        # Augment with assignment information
        result = [self._get_api_consultant_for_base(c, assignments) for c in consultants]

        # Filter on project name
        if result and project_name:
            result = [
                c for c in result
                if any(
                    project_name.lower() in (p.projectName.lower() + p.clientName.lower())
                    for p in c.projects
                )
            ]

        # Filter on available hours
        if result and hours_available:
            try:
                min_hours = int(hours_available)
                result = [
                    c for c in result
                    if (AVAILABLE_HOURS_PER_MONTH * 2 - c.forecastThisMonth - c.forecastNextMonth) >= min_hours
                ]
            except ValueError:
                pass

        return result

    def create_api_consultant(self, consultant: Consultant) -> ApiConsultant:
        consultant_db_service.create_consultant(consultant)
        assignments = assignment_db_service.get_assignments()
        return self._get_api_consultant_for_base(consultant, assignments)

    def _get_api_consultant_for_base(self, consultant: Consultant, assignments: list) -> ApiConsultant:
        api_consultant = ApiConsultant(
            id=consultant.id,
            name=consultant.name,
            email=consultant.email,
            phone=consultant.phone,
            consultantPhotoUrl=consultant.consultantPhotoUrl,
            location=consultant.location,
            skills=consultant.skills,
            certifications=consultant.certifications,
            roles=consultant.roles,
            projects=[],
            forecastThisMonth=0,
            forecastNextMonth=0,
            deliveredLastMonth=0,
            deliveredThisMonth=0,
        )

        consultant_assignments = [a for a in assignments if a.consultantId == consultant.id]

        for assignment in consultant_assignments:
            project = project_db_service.get_project_by_id(assignment.projectId)
            forecast_hours = self._find_hours(assignment.forecast)
            delivered_hours = self._find_hours(assignment.delivered)

            api_consultant.projects.append(ApiConsultantAssignment(
                projectName=project.name,
                projectDescription=project.description,
                projectLocation=project.location,
                clientName=project.clientName,
                clientContact=project.clientContact,
                clientEmail=project.clientEmail,
                role=assignment.role,
                forecastThisMonth=forecast_hours["thisMonth"],
                forecastNextMonth=forecast_hours["nextMonth"],
                deliveredLastMonth=delivered_hours["lastMonth"],
                deliveredThisMonth=delivered_hours["thisMonth"],
            ))
            api_consultant.forecastThisMonth += forecast_hours["thisMonth"]
            api_consultant.forecastNextMonth += forecast_hours["nextMonth"]
            api_consultant.deliveredLastMonth += delivered_hours["lastMonth"]
            api_consultant.deliveredThisMonth += delivered_hours["thisMonth"]

        return api_consultant

    def _find_hours(self, hours: list) -> dict:
        """Extract last, this, and next month's hours from an array of HoursEntry dicts."""
        now = datetime.now()
        this_month = now.month  # 1-based
        this_year = now.year

        last_month = 12 if this_month == 1 else this_month - 1
        last_year = this_year - 1 if this_month == 1 else this_year

        next_month = 1 if this_month == 12 else this_month + 1
        next_year = this_year + 1 if this_month == 12 else this_year

        last_month_hours = 0.0
        this_month_hours = 0.0
        next_month_hours = 0.0

        if hours:
            for h in hours:
                m = h.get("month", 0) if isinstance(h, dict) else h.month
                y = h.get("year", 0) if isinstance(h, dict) else h.year
                hrs = h.get("hours", 0) if isinstance(h, dict) else h.hours
                if m == last_month and y == last_year:
                    last_month_hours = hrs
                elif m == this_month and y == this_year:
                    this_month_hours = hrs
                elif m == next_month and y == next_year:
                    next_month_hours = hrs

        return {
            "lastMonth": last_month_hours,
            "thisMonth": this_month_hours,
            "nextMonth": next_month_hours,
        }

    def charge_time_to_project(self, project_name: str, consultant_id: str, hours: float) -> ApiChargeTimeResponse:
        # Import here to avoid circular imports
        from .project_api_service import project_api_service

        projects = project_api_service.get_api_projects(project_name, "")
        if not projects:
            raise HttpError(404, f"Project not found: {project_name}")
        elif len(projects) > 1:
            raise HttpError(406, f"Multiple projects found with the name: {project_name}")

        project = projects[0]
        month = datetime.now().month
        year = datetime.now().year
        remaining_forecast = assignment_db_service.charge_hours_to_project(
            project.id, consultant_id, month, year, hours
        )

        if remaining_forecast < 0:
            message = (
                f'Charged {hours} hours to {project.clientName} on project "{project.name}". '
                f'You are {-remaining_forecast} hours over your forecast this month.'
            )
        else:
            message = (
                f'Charged {hours} hours to {project.clientName} on project "{project.name}". '
                f'You have {remaining_forecast} hours remaining this month.'
            )

        return ApiChargeTimeResponse(
            clientName=project.clientName,
            projectName=project.name,
            remainingForecast=remaining_forecast,
            message=message,
        )


# Singleton instance
consultant_api_service = ConsultantApiService()

File: samples/da-trey-research-cs/Services/ConsultantApiService.cs
using TreyResearch.Models;

namespace TreyResearch.Services;

public class ConsultantApiService
{
    private static readonly ConsultantApiService _instance = new();
    public static ConsultantApiService Instance => _instance;

    private const int AvailableHoursPerMonth = 160;

    public async Task<ApiConsultant> GetApiConsultantByIdAsync(string consultantId)
    {
        var consultant = await ConsultantDbService.Instance.GetConsultantByIdAsync(consultantId);
        var assignments = await AssignmentDbService.Instance.GetAssignmentsAsync();
        return await GetApiConsultantForBaseConsultantAsync(consultant, assignments);
    }

    public async Task<List<ApiConsultant>> GetApiConsultantsAsync(
        string consultantName, string projectName, string skill,
        string certification, string role, string hoursAvailable)
    {
        var consultants = await ConsultantDbService.Instance.GetConsultantsAsync();
        var assignments = await AssignmentDbService.Instance.GetAssignmentsAsync();

        // Filter on base properties
        if (!string.IsNullOrEmpty(consultantName))
        {
            consultants = consultants.Where(c => c.Name.Contains(consultantName, StringComparison.OrdinalIgnoreCase)).ToList();
        }
        if (!string.IsNullOrEmpty(skill))
        {
            consultants = consultants.Where(c => c.Skills.Any(s => s.Contains(skill, StringComparison.OrdinalIgnoreCase))).ToList();
        }
        if (!string.IsNullOrEmpty(certification))
        {
            consultants = consultants.Where(c => c.Certifications.Any(s => s.Contains(certification, StringComparison.OrdinalIgnoreCase))).ToList();
        }
        if (!string.IsNullOrEmpty(role))
        {
            consultants = consultants.Where(c => c.Roles.Any(s => s.Contains(role, StringComparison.OrdinalIgnoreCase))).ToList();
        }

        // Augment with assignment information
        var result = new List<ApiConsultant>();
        foreach (var c in consultants)
        {
            result.Add(await GetApiConsultantForBaseConsultantAsync(c, assignments));
        }

        // Filter on project name
        if (!string.IsNullOrEmpty(projectName))
        {
            result = result.Where(c => c.Projects.Any(p =>
                (p.ProjectName.ToLower() + p.ClientName.ToLower()).Contains(projectName)
            )).ToList();
        }

        // Filter on available hours
        if (!string.IsNullOrEmpty(hoursAvailable) && int.TryParse(hoursAvailable, out var minHours))
        {
            result = result.Where(c =>
            {
                var available = AvailableHoursPerMonth * 2 - c.ForecastThisMonth - c.ForecastNextMonth;
                return available >= minHours;
            }).ToList();
        }

        return result;
    }

    public async Task<ApiConsultant> CreateApiConsultantAsync(Consultant consultant)
    {
        await ConsultantDbService.Instance.CreateConsultantAsync(consultant);
        var assignments = await AssignmentDbService.Instance.GetAssignmentsAsync();
        return await GetApiConsultantForBaseConsultantAsync(consultant, assignments);
    }

    private async Task<ApiConsultant> GetApiConsultantForBaseConsultantAsync(Consultant consultant, List<Assignment> assignments)
    {
        var result = new ApiConsultant
        {
            Id = consultant.Id,
            Name = consultant.Name,
            Email = consultant.Email,
            Phone = consultant.Phone,
            ConsultantPhotoUrl = consultant.ConsultantPhotoUrl,
            Location = consultant.Location,
            Skills = consultant.Skills,
            Certifications = consultant.Certifications,
            Roles = consultant.Roles
        };

        var consultantAssignments = assignments.Where(a => a.ConsultantId == consultant.Id).ToList();

        foreach (var assignment in consultantAssignments)
        {
            var project = await ProjectDbService.Instance.GetProjectByIdAsync(assignment.ProjectId);
            var (forecastLastMonth, forecastThisMonth, forecastNextMonth) = FindHours(assignment.Forecast);
            var (deliveredLastMonth, deliveredThisMonth, deliveredNextMonth) = FindHours(assignment.Delivered);

            result.Projects.Add(new ApiConsultantAssignment
            {
                ProjectName = project.Name,
                ProjectDescription = project.Description,
                ProjectLocation = project.Location,
                ClientName = project.ClientName,
                ClientContact = project.ClientContact,
                ClientEmail = project.ClientEmail,
                Role = assignment.Role,
                ForecastThisMonth = forecastThisMonth,
                ForecastNextMonth = forecastNextMonth,
                DeliveredLastMonth = deliveredLastMonth,
                DeliveredThisMonth = deliveredThisMonth
            });

            result.ForecastThisMonth += forecastThisMonth;
            result.ForecastNextMonth += forecastNextMonth;
            result.DeliveredLastMonth += deliveredLastMonth;
            result.DeliveredThisMonth += deliveredThisMonth;
        }

        return result;
    }

    private static (double lastMonth, double thisMonth, double nextMonth) FindHours(List<HoursEntry> hours)
    {
        var now = DateTime.Now;
        var thisMonth = now.Month;
        var thisYear = now.Year;

        var lastMonth = thisMonth == 1 ? 12 : thisMonth - 1;
        var lastYear = thisMonth == 1 ? thisYear - 1 : thisYear;

        var nextMonth = thisMonth == 12 ? 1 : thisMonth + 1;
        var nextYear = thisMonth == 12 ? thisYear + 1 : thisYear;

        var lastMonthHours = hours.FirstOrDefault(h => h.Month == lastMonth && h.Year == lastYear)?.Hours ?? 0;
        var thisMonthHours = hours.FirstOrDefault(h => h.Month == thisMonth && h.Year == thisYear)?.Hours ?? 0;
        var nextMonthHours = hours.FirstOrDefault(h => h.Month == nextMonth && h.Year == nextYear)?.Hours ?? 0;

        return (lastMonthHours, thisMonthHours, nextMonthHours);
    }

    public async Task<ApiChargeTimeResponse> ChargeTimeToProjectAsync(string projectName, string consultantId, double hours)
    {
        var projects = await ProjectApiService.Instance.GetApiProjectsAsync(projectName, "");
        if (projects.Count == 0)
        {
            throw new HttpError(404, $"Project not found: {projectName}");
        }
        if (projects.Count > 1)
        {
            throw new HttpError(406, $"Multiple projects found with the name: {projectName}");
        }

        var project = projects[0];
        var month = DateTime.Now.Month;
        var year = DateTime.Now.Year;
        var remainingForecast = await AssignmentDbService.Instance.ChargeHoursToProjectAsync(project.Id, consultantId, month, year, hours);

        string message;
        if (remainingForecast < 0)
        {
            message = $"Charged {hours} hours to {project.ClientName} on project \"{project.Name}\". You are {-remainingForecast} hours over your forecast this month.";
        }
        else
        {
            message = $"Charged {hours} hours to {project.ClientName} on project \"{project.Name}\". You have {remainingForecast} hours remaining this month.";
        }

        return new ApiChargeTimeResponse
        {
            ClientName = project.ClientName,
            ProjectName = project.Name,
            RemainingForecast = remainingForecast,
            Message = message
        };
    }
}
